oap-server/server-library/library-util/src/main/java/org/apache/skywalking/oap/server/library/util/FileChangeMonitor.java

```diff
@@ -0,0 +1,188 @@
+     */
+    private final FileChangedNotifier notifier;
+    /**
+     * The timestamp when last time do status checked.
+     */
+    private long lastCheckTimestamp = 0;
+    /**
+     * The last modify time of the {@link #filePath}
+     */
+    private long lastModifiedTimestamp = 0;
+
+    /**
+     * Start the file monitor for this instance.
+     */
+    public synchronized void start() {
+        if (FILE_MONITOR_TASK_SCHEDULER == null) {
+            FILE_MONITOR_TASK_SCHEDULER = Executors.newSingleThreadScheduledExecutor()
+                                                   .scheduleAtFixedRate(
+                                                       FileChangeMonitor::run, 1, 1,
+                                                       TimeUnit.SECONDS
+                                                   );
+        }
+
+        this.checkAndNotify();
+        MONITOR_INSTANCES.add(this);
+    }
+
+    public synchronized void stop() {
+        MONITOR_INSTANCES.remove(this);
+    }
+
+    /**
+     * Check the file status, if changed, send the notification.
+     */
+    private void checkAndNotify() {
+        if (System.currentTimeMillis() - lastCheckTimestamp < watchingPeriodInSec * 1000) {
+            // Don't reach the period threshold, ignore this check.
+            return;
+        }
+        File targetFile = new File(filePath);
+        if (!targetFile.exists() && acceptFileNotExisting) {
+            notifier.fileNotFound();
+        }
+        if (targetFile.isFile()) {
+            long lastModified = targetFile.lastModified();
+
+            if (lastModified != lastModifiedTimestamp) {
+                try (FileInputStream fileInputStream = new FileInputStream(targetFile)) {
+                    notifier.fileChanged(fileInputStream);
+                } catch (FileNotFoundException e) {
+                    log.error("The existed file turns to missing, watch file=" + filePath, e);
+                } catch (IOException e) {
+                    log.error("Read file failure, watch file=" + filePath, e);
+                } finally {
+                    lastModifiedTimestamp = lastModified;
+                }
+            }
+        }
+    }
+
+    /**
+     * Check all registered file.
+     */
+    private static void run() {
+        MONITOR_INSTANCES.forEach(monitor -> {
+            try {
+                monitor.checkAndNotify();
+            } catch (Throwable e) {
+                log.error("Error happens during monitoring file=" + monitor.filePath, e);
+            }
+        });
+    }
+
+    /**
+     * The callback when file changed.
+     */
+    public interface FileChangedNotifier {
+        /**
+         * Notify the new content by providing the file input stream
+         *
+         * @param readableStream points to the new content
+         * @throws IOException if error happens during reading.
+         */
+        void fileChanged(InputStream readableStream) throws IOException;
```

Comment:
I don't think it's a good design, unless you're perfectly sure that the caller will always use the file content as an `InputStream`, otherwise, you create the USELESS input stream too eagerly

Reply:
What else could they use?